### ISSUE TITLE
Warning: `beets.util.confit is deprecated; use confuse instead`

### ISSUE THREAD

**Issue by Bastes** [adamjakab/BeetsPluginDescribe#4]
### Problem

Using the plugin, I get this warning on every command, before any other output:

```
/homes/my_home/.local/lib/python3.8/site-packages/beets/util/confit.py:21: UserWarning: beets.util.confit is deprecated; use confuse instead
  warnings.warn("beets.util.confit is deprecated; use confuse instead")
```

This warning appeared after I installed and activated the plugin, and does not when the plugin is not enabled nor installed, but on every command I tried (not just `describe`).

The warning does not seem to have any other consequence so far, all commands including `describe` seem to work fine.

### Setup

* OS: Synology DSM 7.0
* Python version: 3.8.8
* Beets version: 1.5.0
* Turning off other plugins made problem go away (yes/no): no
  turning off all other plugins yield the same result as long as `describe` is on
  turning `describe` off removes the warning

My configuration (output of `beet config`) is:

```yaml
directory: ~/music

library: ~/.config/beets/musiclibrary.db

import:
  move: yes

plugins:
  - describe
  - embedart
  - fetchart
  - lastgenre
  - lyrics
  - rewrite
  - the

paths:
  default: %the{$albumartist}/$album/$track $title

albumart:
  ifempty: yes

fetchart:
  minwidth: 300
  maxwidth: 1000
  enforce_ratio: yes

rewrite:
  artist .*johnny clegg.*: Johnny Clegg
  artist .*hans zimmer.*: Hans Zimmer
  artist .*“weird al” yankovic.*: Weird Al Yankovic
```

My plugin version (output of `beet describe -v`) is:

```text
/homes/my_home/.local/lib/python3.8/site-packages/beets/util/confit.py:21: UserWarning: beets.util.confit is deprecated; use confuse instead
  warnings.warn("beets.util.confit is deprecated; use confuse instead")
describe: Describe(beets-describe) plugin for Beets: v0.0.3
```

**Comment by Bastes** [comment]
Also, I :heart: that plugin, very useful to search for artists, composers, etc. written in kanji (since I'm a non japanese speaker that listens to a ton of anime music).

**Comment by janpeeters70** [comment]
I have the same problem, just came here to report it. 
Are you still actively developing this plugin @adamjakab? I find it very useful. 
Thanks, Jan

**Comment by JOJ0** [comment]
Same here, I am on Manjaro Linux, Python 3.9, if that matters, guess not, sounds like a warning from a beets thing itself.

Anyone found a workaround or has the skills to fix this as the warning suggests (whatever confuse and confit are). I would love to do it but I am new to beets and haven't yet started writing plugins. Will do, in the future, at some point, not now! ;-)
@adamjakab by any chance do you thing you will find the time to fix this or at least have a feeling if it's a 5 minute thing to fix or  the change to confuse is a bigger thing. Just curious. Thanks for any thoughs :-)

**Comment by JOJ0** [comment]
Well the most dirty workaround probably is to just comment out line 21 in /homes/my_home/.local/lib/python3.8/site-packages/beets/util/confit.py, haha ;-)

```
...
import warnings
# warnings.warn("beets.util.confit is deprecated; use confuse instead")
...
```

**Comment by Bastes** [comment]
Interesting... so the warning comes from beets itself and not this plugin? Then maybe we should escalate to the beets repository itself, wdyt?

**Comment by Bastes** [comment]
Cancelling and replacing my previous comment: I now think this should not be escalated to beets repo, the problem is that *they* have deprecated use of their `confit.py` file that *this* lib uses, so I guess the problem needs to be addressed from here.

So, the question is: where is it used here? If we can locate it, maybe we can try and disable the place that does and see if everything keeps on working without... or search how to replace it.

Wdyt?

**Comment by janpeeters70** [comment]
There were more reports at https://github.com/beetbox/beets. Look in discussions and issues. I also filled one. 
As far as I can determine, it's just that plugins need to be updated for Beets 1.5 which doesn't use confit.py anymore. There were more plugins that had this problem. So, as annoying as having to update plugins after a major new release might be, it's I guess the result of progress on Beets side and not something they need to fix or revert.

**Comment by Bastes** [comment]
Sure. What I meant was, maybe we can offer a pull request if we know how to adapt to that new normal...

**Comment by Bastes** [comment]
Ok, so, digging a little bit I found this:

> Similarly, we’ve replaced beets’ configuration library (previously called Confit) with a standalone version called Confuse. Where you used to do from beets.util import confit, now just do import confuse. The code is almost identical apart from the name change. Again, we’ll re-export at the old location (with a deprecation warning) for backwards compatibility, but we might stop doing this in a future release.

(https://beets.readthedocs.io/en/stable/changelog.html)

So, this looks rather straightforward... but I don't know how to test it locally. So, I'll fork and branch and offer and MR, and if one of us can tell me how to test it and/or want to test it they can.

Stay tuned.

**Comment by Bastes** [comment]
Ok, so, here's a branch that might fix it, but again I don't know how to test it. If someone is wiser in the ways of python plugins development for beets can tell me how to do that, I'll happily do, or if you can and want to test it yourself knock yourself out. I'd just prefer not to make an MR that doesn't work:
https://github.com/Bastes/BeetsPluginDescribe/tree/confit-to-confuse

**Comment by JOJ0** [comment]
You mean test as in unittest or test as in try out if it works in your own beets installation?

**Comment by Bastes** [comment]
The latter should be plenty sufficient :) since it's a change of dependency, either it works or it'll bug I think? (I'm not very used to python)

**Comment by JOJ0** [comment]
Hi, I can confirm that it works :-))

```
$ grep warning beets/util/confit.py                                                                                                                                                              
import warnings
warnings.warn("beets.util.confit is deprecated; use confuse instead")
del key, value, warnings, confuse


$ pip uninstall beets-describe                                                                                                                                                                 
Found existing installation: beets-describe 0.0.4
Uninstalling beets-describe-0.0.4:
  Would remove:
    /home/jojo/.local/lib/python3.10/site-packages/beets_describe-0.0.4-py3.10.egg-info
    /home/jojo/.local/lib/python3.10/site-packages/beetsplug/describe/*
Proceed (y/n)? 




$ pip install --user git+https://github.com/Bastes/BeetsPluginDescribe@confit-to-confuse                                                                                                 
Collecting git+https://github.com/Bastes/BeetsPluginDescribe@confit-to-confuse
  Cloning https://github.com/Bastes/BeetsPluginDescribe (to revision confit-to-confuse) to /tmp/pip-req-build-n2y9_0li
  Running command git clone -q https://github.com/Bastes/BeetsPluginDescribe /tmp/pip-req-build-n2y9_0li
  Running command git checkout -b confit-to-confuse --track origin/confit-to-confuse
  Switched to a new branch 'confit-to-confuse'
  Branch 'confit-to-confuse' set up to track remote branch 'confit-to-confuse' from 'origin'.
Requirement already satisfied: beets>=1.4.9 in /home/jojo/git/beets (from beets-describe==0.0.4) (1.6.1)
Requirement already satisfied: numpy in /home/jojo/.local/lib/python3.10/site-packages (from beets-describe==0.0.4) (1.22.0)
Requirement already satisfied: pandas in /usr/lib/python3.10/site-packages (from beets-describe==0.0.4) (1.3.5)
Requirement already satisfied: termtables in /usr/lib/python3.10/site-packages (from beets-describe==0.0.4) (0.2.4)
Requirement already satisfied: termplotlib in /usr/lib/python3.10/site-packages (from beets-describe==0.0.4) (0.3.9)
Requirement already satisfied: unidecode in /usr/lib/python3.10/site-packages (from beets>=1.4.9->beets-describe==0.0.4) (1.3.2)
Requirement already satisfied: musicbrainzngs>=0.4 in /usr/lib/python3.10/site-packages (from beets>=1.4.9->beets-describe==0.0.4) (0.7.1)
Requirement already satisfied: pyyaml in /usr/lib/python3.10/site-packages (from beets>=1.4.9->beets-describe==0.0.4) (5.4.1)
Requirement already satisfied: mediafile>=0.9.0 in /usr/lib/python3.10/site-packages (from beets>=1.4.9->beets-describe==0.0.4) (0.9.0)
Requirement already satisfied: confuse>=1.5.0 in /usr/lib/python3.10/site-packages (from beets>=1.4.9->beets-describe==0.0.4) (1.6.0)
Requirement already satisfied: munkres>=1.0.0 in /usr/lib/python3.10/site-packages (from beets>=1.4.9->beets-describe==0.0.4) (1.1.4)
Requirement already satisfied: jellyfish in /usr/lib/python3.10/site-packages (from beets>=1.4.9->beets-describe==0.0.4) (0.8.9)
Requirement already satisfied: mutagen>=1.45 in /usr/lib/python3.10/site-packages (from mediafile>=0.9.0->beets>=1.4.9->beets-describe==0.0.4) (1.45.1)
Requirement already satisfied: six>=1.9 in /usr/lib/python3.10/site-packages (from mediafile>=0.9.0->beets>=1.4.9->beets-describe==0.0.4) (1.16.0)
Requirement already satisfied: python-dateutil>=2.7.3 in /usr/lib/python3.10/site-packages (from pandas->beets-describe==0.0.4) (2.8.2)
Requirement already satisfied: pytz>=2017.3 in /usr/lib/python3.10/site-packages (from pandas->beets-describe==0.0.4) (2021.3)
Using legacy 'setup.py install' for beets-describe, since package 'wheel' is not installed.
Installing collected packages: beets-describe
    Running setup.py install for beets-describe ... done
Successfully installed beets-describe-0.0.4

**Comment by Bastes** [comment]
Well, that looks ready to be reviewed :wink: I'll send the PR.

**Comment by Bastes** [comment]
Here's the PR :slightly_smiling_face: https://github.com/adamjakab/BeetsPluginDescribe/pull/5

**Comment by janpeeters70** [comment]
Thanks a lot for fixing this. @adamjakab Has another great plugin that has the same problem:

**Comment by janpeeters70** [comment]
One additional question… I'm not too acquainted with PR's. Will the PR have to be processed by the Beets developers or by @adamjakab To be included in the plugin when you install it the regular way? And if it is the Beets developers will they need to be informed about the PR or will this go automatically.

**Comment by Bastes** [comment]
@janpeeters70 to answer your questions in order:

* Sure, but I won't have much time right now ; you can also look at the diffs from my PR and apply them in that other plugin, the idea is simple: instead of importing stuff from `beets.util.confit` you want to import them from `confuse` instead. So:
`from beets.util.confit import ConfigSource, load_yaml`
becomes:
`from confused import ConfigSource, load_yaml`
(for example)

* And yes, the PR being Pull *requests*, they're just requests that maintainer(s) of a repository have to accept to get merged in... which means in our case @adamjakab unless he configured this repository to be administered by someone else. I hope he's listening to what's happening on his repository but not knowing him personally I can't do better than write here to get his attention I'm afraid :sweat_smile: not knowing much about the python ecosystem, I guess it'd be possible for beets to change the plugin's source but I wouldn't position myself as maintainer anyways so you'd have to find someone else to make this a long-term solution in case @adamjakab does not answer or something.

So I hope this answers your questions properly :slightly_smiling_face: I'll see what I can do for your other plugin if I find the time.

**Comment by janpeeters70** [comment]
Hi Michel (@Bastes), this answers my questions really well. Thank you very much!
I think the comparing would work, but I would have to dive into how to make a new version/branch? and pull request to be able to contribute. Thanks for having a look when you have time. I'll try to do the same.

**Comment by Bastes** [comment]
@janpeeters70 Making a new version/branch is as simple as clicking on the "fork" button on the top right. Then you don't really need to make a new branch, you can work on then `devel` branch in your fork if you like. As for the pull request, it's as simple as going to the "Pull requests" tab and clicking on the "New pull request" button. Nothing to be afraid of really, and usually the maintainers of a project love a good pull request (less work for them).

But then again, it's ok if you don't feel like doing it :wink:

**Comment by janpeeters70** [comment]
That sounds not too difficult. Thanks for the clarification! I have a look when I find the time.

**Comment by JOJ0** [comment]
@janpeeters70 The goingrunning plugin is maintained in adamjakab's personal git repo as you can see in the plugins git url. The beets maintainers do not have access to any third party repos.

There _are_ beets plugins that are officially shipping together with the installation of beets. Out of the box plugins so to say. Have a look into the beetsplug folder _inside_ of beetbox/beets repo. Those are maintained by beets devs themselves.

**Comment by janpeeters70** [comment]
Thanks for that clarification @JOJ0. So if I understand well, I can then also install the plugin directly from @Bastes repo until @adamjakab has integrated his pull request into the master branch?

**Comment by Bastes** [comment]
Well, sure that should work?

**Comment by Bastes** [comment]
But apart from the warning @janpeeters70 as long as the deprecated lib is not removed the plugin will continue to work as intended.

**Comment by janpeeters70** [comment]
> Well, sure that should work?

@Bastes I don't understand what you are writing. I reads like a confirmation, but then there is the question mark?

Thanks for the notion about the working of the plugin. I somehow find the warning annoying to see it every time. So I'm going to try to get rid of it.

**Comment by Bastes** [comment]
@janpeeters70 It's just because I haven't tested it myself yet :wink:

**Comment by janpeeters70** [comment]
I tried:
`$ pip uninstall beets-describe`
And after that
`pip install --user git+https://github.com/Bastes/BeetsPluginDescribe@confit-to-confuse`

But I still get the deprecated error:
```
$ beet -v describe genre
user configuration: /Users/jpeeters/.config/beets/config.yaml
data directory: /Users/jpeeters/.config/beets
plugin paths: 
/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/beets/util/confit.py:19: UserWarning: beets.util.confit is deprecated; use confuse instead
  warnings.warn("beets.util.confit is deprecated; use confuse instead")
```
Maybe something wrong on my side but message still occurs.
The output it generates is correct though so plugin works.

**Comment by JOJ0** [comment]
The /opt/local/Library/... path tells you that you had installed the plugin systemwide.

you now seem to have installed them in two places: systemwide and also in your userhome (that's what --local in setup command did)

try to uninstall syswide version with:

sudo pip uninstall beets-describe

HTH

**Comment by JOJ0** [comment]
> Thanks a lot for fixing this. @adamjakab Has another great plugin that has the same problem: 
> 
> [https://github.com/adamjakab/BeetsPluginGoingRunning](https://github.com/adamjakab/BeetsPluginGoingRunning)

BTW someone fixed it 20 days ago already. PR pending...

**Comment by janpeeters70** [comment]
> The /opt/local/Library/... path tells you that you had installed the plugin systemwide.
> 
> you now seem to have installed them in two places: systemwide and also in your userhome (that's what --local in setup command did)
> 
> try to uninstall both versions with:
> 
> sudo pip uninstall beets-describe
> 
> and
> 
> pip uninstall beets-describe
> 
> Check if it's gone with
> 
> pip list -v
> 
> Really check all your dirs in $PYTHONPATH if beets-describe package is gone. Use find and grep for that.
> 
> Then reinstall using the pip install --user command
> 
> HTH

Thanks for your keen eye @JOJ0 I indeed have it installed system wide via Macports. Didn't realize that. I'll have to figure out if I can use a systemwide Beets via Macports via a pip installed describe.

**Comment by JOJ0** [comment]
you can always configure your plugin paths manually in beets config.yaml

I'll send you example of my conf when not AFK ;-)

**Comment by janpeeters70** [comment]
> you can always configure your plugin paths manually in beets config.yaml
> 
> I'll send you example of my conf when not AFK ;-)

That would be great @JOJ0
Thanks a lot for all your help so far.

**Comment by JOJ0** [comment]
Hi,
I don't know how many plugins you have installed but if it is just a few that are not "beets out of the box" plugins then I suggest you configure your plugin search paths manually and install all your plugins from git directly.

This is what i have in my beets config.

```
pluginpath:
- ~/.local/lib/python3.9/site-packages/beetsplug  # -> Everything you installed with pip install --local
- ~/git/beets/beetsplug  # -> beets OOTB plugins. You see I even installed beets from git repo. It's a very well maintained project and you don't need to wait for releases. Once a pull request is merged, you just `git pull` and run the latest verstion from master. Not sure what macports does but probably they rely on a released version, right?
- ~/git/mried_beetsplug/beetsplug  #  -> 3rd party plugin directly running from git, no setup.py file
- ~/git/whatlastgenre/plugin/beets/beetsplug  #  -> 3rd party plugin directly running from git, no setup.py file
```


- To install into user home site-packages (local) dir you know the command already. If you remove the --local and add sudo in front of the pip command you would install to your systemwide dir. Don't do that, it's only useful in multiuser environments and complicates stuff if you sometimes install this way or that way
- You know already how to clone a plugin repo from github to your disk I suppose.
- Some plugins come with a setup.py file in the repo. Use that! Usually using `python3 setup.py install`. That installs to .... I am not sure where but I think to a systemwide dir. Also sometimes executables want to be put into /usr/local/bin, thus you get permission errors and need to use sudo with these kinds of installations anyway. Sorry no better answer here...

Always use `pip list -v` after each step to really see where a `pip install` or `python setup.py install` has put your packages files!!!

**Comment by JOJ0** [comment]
Here an example on how to use pip install to user dir with a library I know well, not a beets plug, but any python package would do here to proof the concept:

```
    ~  pip list -v | grep discogs                                                                           ✔ 
python3-discogs-client 2.3.12      /home/jojo/.local/lib/python3.10/site-packages
    ~                                                                                                       ✔ 
    ~                                                                                                       ✔ 
    ~  pip uninstall python3-discogs-client                                                                 ✔ 
Found existing installation: python3-discogs-client 2.3.12
Uninstalling python3-discogs-client-2.3.12:
  Would remove:
    /home/jojo/.local/lib/python3.10/site-packages/discogs_client/*
    /home/jojo/.local/lib/python3.10/site-packages/python3_discogs_client-2.3.12-py3.10.egg-info
Proceed (y/n)? y
  Successfully uninstalled python3-discogs-client-2.3.12
    ~                                                                                                       ✔ 
    ~                                                                                                       ✔ 
    ~                                                                                                       ✔ 
    ~                                                                                                       ✔ 
    ~  cd git                                                                                               ✔ 
    ~/git  git clone git@github.com:joalla/discogs_client.git                                               ✔ 
Cloning into 'discogs_client'...
Enter passphrase for key '/home/jojo/.ssh/id_ed25519': 
remote: Enumerating objects: 1267, done.
remote: Counting objects: 100% (224/224), done.
remote: Compressing objects: 100% (103/103), done.
remote: Total 1267 (delta 129), reused 190 (delta 116), pack-reused 1043
Receiving objects: 100% (1267/1267), 280.70 KiB | 1.65 MiB/s, done.
Resolving deltas: 100% (720/720), done.
    ~/git                                                                                            ✔  5s  
    ~/git                                                                                                   ✔ 
    ~/git  cd discogs_client                                                                                ✔ 
    ~/git/discogs_client    master                                                                       ✔ 
    ~/git/discogs_client    master  ls setup.py                                                          ✔ 
setup.py
    ~/git/discogs_client    master                                                                       ✔ 
    ~/git/discogs_client    master  pip install --user ./                                                ✔ 
Processing /home/jojo/git/discogs_client
Requirement already satisfied: requests in /usr/lib/python3.10/site-packages (from python3-discogs-client==2.3.12) (2.26.0)
Requirement already satisfied: oauthlib in /usr/lib/python3.10/site-packages (from python3-discogs-client==2.3.12) (3.1.1)
Requirement already satisfied: python-dateutil in /usr/lib/python3.10/site-packages (from python3-discogs-client==2.3.12) (2.8.2)
Requirement already satisfied: six>=1.5 in /usr/lib/python3.10/site-packages (from python-dateutil->python3-discogs-client==2.3.12) (1.16.0)
Requirement already satisfied: chardet>=3.0.2 in /usr/lib/python3.10/site-packages (from requests->python3-discogs-client==2.3.12) (4.0.0)
Requirement already satisfied: idna>=2.5 in /usr/lib/python3.10/site-packages (from requests->python3-discogs-client==2.3.12) (3.3)
Requirement already satisfied: urllib3>=1.21.1 in /usr/lib/python3.10/site-packages (from requests->python3-discogs-client==2.3.12) (1.26.7)
Using legacy 'setup.py install' for python3-discogs-client, since package 'wheel' is not installed.
Installing collected packages: python3-discogs-client
    Running setup.py install for python3-discogs-client ... done
Successfully installed python3-discogs-client-2.3.12
    ~/git/discogs_client    master                                                                       ✔ 
    ~/git/discogs_client    master                                                                       ✔ 
    ~/git/discogs_client    master                                                                       ✔ 
    ~/git/discogs_client    master                                                                       ✔ 
    ~/git/discogs_client    master  pip list -v | grep discogs                                           ✔ 
python3-discogs-client 2.3.12      /home/jojo/.local/lib/python3.10/site-packages
    ~/git/discogs_client    master                                                                       ✔ 
    ~/git/discogs_client    master                                                                       ✔ 
    ~/git/discogs_client    master                                                                       ✔ 
    ~/git/discogs_client    master                                                                       ✔ 
    ~/git/discogs_client    master 

**Comment by janpeeters70** [comment]
Hi @JOJ0 thank you so much for this very thorough explanation. Really kind to take the time for this! I don't know yet how to clone a repo from Github. And where would be a good place to put the Beets repo (in macOS). It sounds like a very good idea to be 'live' connected to this repo. So I have to do some research for that. I'll do this in a few days because I'm currently quite busy. After that I'll report back. Thanks again.

**Comment by janpeeters70** [comment]
@JOJ0, I made it work thanks to your instructions. Cloning a Git on macOS wasn't as difficult as I feared, from there using the commands you provided worked like a charm.

**Comment by JOJ0** [comment]
one thing on a quick look: you need to clone the fixed repo and "git checkout" the fixed feature branch, not the original one. when you clone you land in master branch by default.

**Comment by janpeeters70** [comment]
Thanks @JOJ0 that was a great pointer. Your feedback and shell print-outs have been instrumental to my understanding how to install beets and plugins directly from Github repo's so thanks a lot!
